Mathlib/Algebra/FreeMonoid/Basic.lean

```diff
@@ -106,6 +107,96 @@ theorem toList_of_mul (x : α) (xs : FreeMonoid α) : toList (of x * xs) = x ::
 @[to_additive]
 theorem of_injective : Function.Injective (@of α) := List.singleton_injective
 
+/-! ### Length -/
+
+section Length
+variable {a : FreeMonoid α}
+
+/-- The length of a free monoid element: 1.length = 0 and (a * b).length = a.length + b.length -/
+@[to_additive "The length of an additive free monoid element: 1.length = 0 and (a + b).length =
+  a.length + b.length"]
+def length (a : FreeMonoid α) : ℕ := List.length a
+
+@[to_additive (attr := simp)]
+theorem length_one : length (1 : FreeMonoid α) = 0 := rfl
+
+@[to_additive (attr := simp)]
+theorem length_eq_zero : length a = 0 ↔ a = 1 := List.length_eq_zero
+
+@[to_additive (attr := simp)]
+theorem length_of (m : α) : length (of m) = 1 := rfl
+
+@[to_additive existing]
+theorem length_eq_one : length a = 1 ↔ ∃ m, a = FreeMonoid.of m :=
+  List.length_eq_one
+
+@[to_additive]
+theorem length_eq_two {v : FreeMonoid α} :
+    v.length = 2 ↔ ∃ c d, v = FreeMonoid.of c * FreeMonoid.of d := List.length_eq_two
+
+@[to_additive (attr := simp)]
+theorem length_mul (a b : FreeMonoid α) : (a * b).length = a.length + b.length :=
+  List.length_append _ _
+
+@[to_additive]
+theorem of_ne_one (a : α) : of a ≠ 1 := by
+  intro h
+  have := congrArg FreeMonoid.length h
+  simp only [length_of, length_one, Nat.succ_ne_self] at this
+
+end Length
+
+/-! ### symbols -/
+
+section Symbols
+variable [DecidableEq α]
+
+/-- The set of unique symbols in a free monoid element -/
+@[to_additive "The set of unique symbols in an additive free monoid element"]
+def symbols (a : FreeMonoid α) : Finset α := List.toFinset a
```

Comment:
Would you mind moving this material to a new file? Because the new import adds 6% to the size of the transitive dependencies of this file.

Reply:
Do you have a good motivation for this new file besides the imports? The extra imports are all just finiteness to get to `Finset`, and I don't think we should be considering `Finset` as something that's heavy to import. If it is, then we should fix it rather than artificially breaking down downstream files.

Reply:
I think it doesn't hurt to separate the basics of the algebraic hierarchy from the finiteness hierarchy.
And I don't think the breaking down has to be very artificial.

Reply:
Okay, but `FreeMonoid` is *literally* `List`, which belongs to the finiteness hierarchy

Reply:
I've removed the material on 'symbols' from this file -- I will soon make another PR with that material in a separate file, Algebra/FreeMonoid/Symbols.lean